server/src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -224,31 +223,47 @@ private static <K, T, H> MapDiff<K, T, H> diffOrEmpty(MapDiff<K, T, H> diff) {
      */
     private static class JdkMapDiff<K, T> extends MapDiff<K, T, Map<K, T>> {
 
+        private JdkMapDiff(
+            KeySerializer<K> keySerializer,
+            ValueSerializer<K, T> valueSerializer,
+            List<K> deletes,
+            List<Map.Entry<K, Diff<T>>> diffs,
+            List<Map.Entry<K, T>> upserts
+        ) {
+            super(keySerializer, valueSerializer, deletes, diffs, upserts);
+        }
+
         private JdkMapDiff(StreamInput in, KeySerializer<K> keySerializer, ValueSerializer<K, T> valueSerializer) throws IOException {
             super(in, keySerializer, valueSerializer);
         }
 
-        private JdkMapDiff(Map<K, T> before, Map<K, T> after, KeySerializer<K> keySerializer, ValueSerializer<K, T> valueSerializer) {
-            super(keySerializer, valueSerializer);
+        private static <K, T> JdkMapDiff<K, T> create(
+            Map<K, T> before,
+            Map<K, T> after,
+            KeySerializer<K> keySerializer,
+            ValueSerializer<K, T> valueSerializer
+        ) {
             assert after != null && before != null;
 
             int inserts = 0;
-            for (Map.Entry<K, T> partIter : after.entrySet()) {
-                final K key = partIter.getKey();
-                T beforePart = before.get(key);
-                if (beforePart == null) {
-                    upserts.put(key, partIter.getValue());
+            var upserts = new ArrayList<Map.Entry<K, T>>();
+            var diffs = new ArrayList<Map.Entry<K, Diff<T>>>();
```

Comment:
I wonder if we could come up with reasonable predictions for initial sizes? @original-brownbear maybe you have any suggestions?

Reply:
I don't think so. Best you could maybe do is some upper bound but that's only really helpful (and to a very minor degree only as well) in case you have less than the default size of 10 in both maps.
Probably not worth it :)